std.crypto.aegis: AEGIS as a stream cipher.

ErikSchlyter opened the pull request:
This patch will enable AEGIS as a stream cipher.

## Background

I need to use a stream cipher in a project I'm working on, and the current available options in the standard library is to use ChaCha/Salsa. I would greatly prefer to use AEGIS as a stream cipher due to massive performance increase, and the fact that a leaked state doesn't leak the key. This brings a very valuable security feature in that a compromised host in my system cannot use the state to determine previous traffic (forward secrecy). If I use a counter based cipher (ChaCha), I need to continuously store the key and the counter as long as the stream is active, which is undesired.

The current implementation of AEGIS in the standard library doesn't support stream mode in the practical sense, since I need to generate the key stream bit by bit over long periods of time, thus I cannot simply encrypt a string of zeros using the `encrypt`-function. The type functions `State128X` and `State256X` are not public, so I cannot reuse them directly.

My project has a strict _no 3rd party dependencies_ policy beyond Zig and its standard library, so the current workaround for the non-public state structs is to simply copy `aegis.zig` from the standard library (along with the license) into my own code, cut away the parts I don't need, and then add a struct that holds the state that can generate key stream continuously. I would greatly prefer if this feature landed in the standard library though.

According to the [IETF draft](https://datatracker.ietf.org/doc/draft-irtf-cfrg-aegis-aead/), all AEGIS variants can be used as stream ciphers.

## Notes on the implementation

`AegisStream` returns a struct with a function `stream` with a similar interface as the one in ChaCha (without the counter, obviously). It is intended to be used for generating streams of arbitrary lengths that doesn't necessarily conform the the stream block size.

I've exposed a function `block` that will generate an individual block, which can be used when generating segments of the key stream over time.

I currently don't know if these AEGIS stream variants (`Aegis128X4Stream`, `Aegis128X2Stream`, `Aegis128LStream`, `Aegis256X4Stream`, `Aegis256X2Stream`, `Aegis256Stream`) should be added to the `pub const stream` in `lib/std/crypto.zig` or not. In some sense it would feel pretty neat, but on the other hand their interface is somewhat different. The `AegisStream` currently doesn't have a `xor`-function for encryption/decryption automatically, so the user would have to do that manually.

I don't have any strong opinions about anything in this patch, except that I kind of need it :) I've tried to mimic the style, structure and tests from the surrounding code. Any feedback is greatly welcomed.

jedisct1 commented:
`stream()` (and `xor()` as well) can simply be implemented as methods of `Aegis128XGeneric` and `Aegis256XGeneric`.

In the specification, for clarity and simplicity, the stream cipher mode is described as equivalent to encrypting all-zero blocks.

The `Update` function is always called with zero-valued message blocks, making the XOR operations effectively no-ops.

Therefore, in a real implementation, it may be more efficient to call a function that directly returns the keystream, rather than using the `Enc` function. This approach is likely to improve performance.

jedisct1 commented:
If you need to generate an practically infinite stream from a single `(key,nonce)` pair (so the same state keeps being reused), rather than public a `block` function, that can be a `std.Random` interface.

ErikSchlyter commented:
Hi! Thanks for the feedback!

The problem with using `Aegis128XGeneric` and `Aegis256XGeneric` is that they don't store the state in the struct. The state is initialized and dismissed inside the functions `encrypt` and `decrypt` as a one-time operation, and I need to hold on to the state between consecutive calls to generate new random bytes.

This is perhaps an undesirable property of a generic stream cipher? (i.e., holding the state might open the door for misuse, etc)

My use-case (perhaps kind of peculiar) is not that I need to generate an infinite stream per se. I need to generate a few blocks now, then separately encrypt the entire state into cold storage (figuratively buried under a public library or something), and then generate a few more blocks a couple of decades later :) Some of the streams need to generate blocks quite often though, and in fast pace.

The alternative is to use a hash ratchet, but then, AEGIS is *fast*, and I don't really need the collision resistance, just the unpredictability.

I have updated the patch with some changes, which is that I added `stream()` to `State128X` and `State256X`, which is basically a combination of `enc()`/`update()` without the superfluous XOR-operations that don't have any effect. I also moved the `stream()` function from `AegisStream` into `Aegis128XGeneric` and `Aegis256XGeneric` (and renamed `block()` into `next()`, for now), so it should now be possible to generate streams using the standard interface.

This still doesn't solve my problem though.

Regarding `std.Random`; moving the `AegisStream` implementation of this patch into `std.Random` and adapt the `block()` function (now `next()`) to fit into `fillFn` would work for my use case, since that seems like a more suitable place for a implementation that stores state. The problem is that `std.Random` doesn't have access to `State128X` or `State256X`, which basically sets me back to square one.

Is this problem solvable, you think?

ErikSchlyter commented:
I'll be looking into it. Thanks!
